### PR TITLE
make ImageDataGenerator behaviour fully seedable/repeatable

### PR DESCRIPTION
**Pull request by wassname**
This makes ImageDataGenerator fully seedable with minimal changes.
- the seed argument in fit is now used
- the seed argument in flow and flow_from_directory now affects
transforms
- added example to docs of transforming images and masks together
- added test of using two seeded streams to get the same result

Related issue: [3059](https://github.com/fchollet/keras/issues/3059)

Use cases: 
- Needing repeatable behaviour from ImageDataGenerator
- Wanting to transform images and masks at the same time, e.g. kaggle ultrasound nerve segmentation challenge

- [x] Tests pass in python2.7 with tensorflow and theano and python3.4 with theano
- [x] pep8 passes
- [x] docs updated

### CODE REVIEW

File: docs/templates/preprocessing/image.md [hunk 1]
```diff
@@ -151,3 +153,38 @@ model.fit_generator(
...
+# we create two instances with the same arguments
+data_gen_args = dict(
+    featurewise_center=True,
+    featurewise_std_normalization=True,
+    rotation_range=90.,
+    width_shift_range=0.1,
+    height_shift_range=0.1,
+    zoom_range=0.2
+)
+generator_i = ImageDataGenerator(**data_gen_args)
```
Comment (fchollet) [inline docs/templates/preprocessing/image.md]: Use explicit variable names in examples (e.g. `image_generator`)

File: docs/templates/preprocessing/image.md [hunk 2]
```diff
@@ -151,3 +153,38 @@ model.fit_generator(
...
+generator_m = ImageDataGenerator(**data_gen_args)
+
+# We need to provide the same seed to the fit and flow methods
+seed = 1
+fake_y = np.zeroes(images.shape)
```
Comment (fchollet) [inline docs/templates/preprocessing/image.md]: If you don't need labels, you should configure the generator not to use labels (`class_mode=None`).

File: docs/templates/preprocessing/image.md [hunk 3]
```diff
@@ -151,3 +153,38 @@ model.fit_generator(
...
+generator_i.fit(images, augment=True, seed=seed)
+generator_m.fit(masks, augment=True, seed=seed)
+flow_images = generator_i.flow(images, fake_y, batch_size=2, seed=seed)
+flow_masks = generator_m.flow(masks, fake_y, batch_size=2, seed=seed)
+
+# combine the two generators into one which gives the first from each
+def dual_gen(flow_images,flow_masks):
+    for [X,_],[y,_] in zip(flow_images,flow_masks):
```
Comment (fchollet) [inline docs/templates/preprocessing/image.md]: Same here, would be much more elegant without the fake labels.

File: docs/templates/preprocessing/image.md [hunk 4]
```diff
@@ -151,3 +153,38 @@ model.fit_generator(
...
+        yield X,y
```
Comment (fchollet) [inline docs/templates/preprocessing/image.md]: PEP8

File: docs/templates/preprocessing/image.md [hunk 5]
```diff
@@ -151,3 +153,38 @@ model.fit_generator(
...
+
+model.fit_generator(
+        dual_gen(flow_images,flow_masks),
```
Comment (fchollet) [inline docs/templates/preprocessing/image.md]: PEP8

File: tests/keras/preprocessing/test_image.py [hunk 6]
```diff
@@ -90,6 +90,55 @@ def test_img_flip(self):
                 assert ((potentially_flipped_x == x).all() or
                         (potentially_flipped_x == flip_axis(x, col_index)).all())
 
+    def test_dual_generators(self):
```
Comment (fchollet) [inline tests/keras/preprocessing/test_image.py]: I don't think this test is useful; please remove.

Comment (wassname) [inline tests/keras/preprocessing/test_image.py]: My logic is that this test will break if any future PR break's repeatability, if you have time could you expand on why you don't think that's worth a test? In the mean time I'll remove it.

**Comment by wassname** [comment]
Thanks for the feedback, I'll make those changes shortly

**Comment by fchollet** [comment]
LGTM, thanks

**Comment by yesufeng** [comment]
does this requires the transformations passed to ImageDataGenerator to be exactly the same for image and mask?

**Comment by wassname** [comment]
If our goal is to transform the image and mask together, then yes, and that's usually what we want. Don't we want the mask to correspond to the same pixels in the image even after transformation. Unless I'm missing something.

See the last example [in the docs](https://keras.io/preprocessing/image/) where they are transformed together.

**Comment by yesufeng** [comment]
yes, I saw that example. Let me give a concrete example, if the images are rescaled but not the mask, e.g., imageGen = ImageDataGenerator(rescale = 1/1000., cval = -1000., fill_mode = 'constant', rotation_range=180.)
                   maskGen = ImageDataGenerator(cval=0., fill_mode='constant', rotation_range=180.), 
these two generators, when we call flow with the same seed and other keyword arguments, then zip them to a final generator, is it guaranteed that these two are synchronized?

**Comment by wassname** [comment]
In your example they wont be.

But instead of taking my word, you can easily test this, and other combinations yourself. Put the same image through both generators and compare them visually or with `assert (a==b).all()`. [Example](https://gist.github.com/wassname/dbb83c6c36dfd249f68b38f1e2118bd2)

Let's say I channel shift the image and not the mask. This change will propagate to other transformations, in this case the vertical flip, and the image and mask will not be synchronized.

**Comment by yesufeng** [comment]
thank you for the explanation! Yeah, feed in the same images will nail it down. It might be desirable to allow things such as rescale to be different for mask vs image without affecting the synchronized generator, for obvious reason.

**Comment by wassname** [comment]
No worries. Yeah agreed.

**Comment by wassname** [comment]
@mositemp I made a quick fix [here](https://github.com/keras-team/keras/pull/7613) but we decided not to merge it. 

Good info on "channel_shift_range" thanks for sharing.
